Sources/SwiftBSON/BSONDocument.swift

```diff
@@ -103,16 +88,13 @@ public struct BSONDocument {
      */
     public init(fromBSON bson: ByteBuffer) throws {
         let storage = BSONDocumentStorage(bson)
-        try storage.validate()
-        self = BSONDocument(fromUnsafeBSON: storage)
+        let keys = try storage.validateAndRetrieveKeys()
+        self = BSONDocument(fromUnsafeBSON: storage, keys: keys)
     }
 
-    private init(fromUnsafeBSON storage: BSONDocumentStorage) {
-        self.keySet = Set()
+    private init(fromUnsafeBSON storage: BSONDocumentStorage, keys: Set<String>) {
```

Comment:
we have always already inspected the keys by the time we call this method, so we can save ourselves iterating through again by passing them in here